### ISSUE TITLE
Connect client connection and heal client monitor stream fail in different times

### ISSUE THREAD

**Issue by Bolodya1997** [networkservicemesh/sdk#925]
## Intro

We have connect client with a gRPC connection to the remote side. We have heal client with a monitor gRPC stream over this connection.

## Expected Behavior

1. If monitor gRPC stream closes, gRPC connection should be closed or be already closed.
2. If monitor gRPC stream fails to start, gRPC connection should be closed and all Requests should return an error.

## Current Behavior

1. Sometimes we have gRPC stream closed, but gRPC connection still working and passing Requests.
2. Sometimes gRPC stream fails to start, but Requests return success.

## Steps to Reproduce

1. Run multiple times heal Endpoint tests.

## Failure logs

```
2021-05-14T11:06:05.4654511Z 2021/05/14 11:04:05 [DEBUG] --- START: TestNSMGR_HealEndpoint/Remote_New
2021-05-14T11:06:05.4655234Z 2021/05/14 11:04:05 [INFO]  Started listening Registry on: tcp://127.0.0.1:36973
2021-05-14T11:06:05.4656499Z 2021/05/14 11:04:05 [INFO]  Started listening NSMgr nsmgr-263cb81b-eebf-4eef-b3bf-87c8f6d111ee on tcp://127.0.0.1:45251
2021-05-14T11:06:05.4658517Z 2021/05/14 11:04:05 [INFO]  Started listening endpoint 7556d6a6-d0f5-49c6-9aec-90d9e47ffe97-forwarder-722ddefe-9a47-4333-ad16-668898ca774b#interpose-nse on tcp://127.0.0.1:37911
2021-05-14T11:06:05.4660415Z 2021/05/14 11:04:05 [INFO]  Started listening NSMgr nsmgr-beb8804b-e1d0-4986-8262-d311316ee98b on tcp://127.0.0.1:44327
2021-05-14T11:06:05.4662452Z 2021/05/14 11:04:05 [INFO]  Started listening endpoint baf61bc0-1a6b-44f2-a0b3-488b83cd4cd2-forwarder-bebfdb01-0f58-4810-8497-eda567dbcbfc#interpose-nse on tcp://127.0.0.1:36087
2021-05-14T11:06:05.4664415Z 2021/05/14 11:04:05 [INFO]  Started listening endpoint 890dc25b-8749-4522-8461-2e4524f293f7-final-endpoint on tcp://127.0.0.1:43643
2021-05-14T11:06:05.4665646Z 2021/05/14 11:04:05 [ERROR] [fs.WatchFile:/var/lib/networkservicemesh/config/excluded_prefixes.yaml] can not create directory: mkdir /var/lib/networkservicemesh: permission denied
2021-05-14T11:06:05.4666617Z 2021/05/14 11:04:05 [DEBUG] tcp://127.0.0.1:37911 READY
2021-05-14T11:06:05.4667101Z 2021/05/14 11:04:05 [DEBUG] tcp://127.0.0.1:45251 READY
2021-05-14T11:06:05.4667570Z 2021/05/14 11:04:05 [DEBUG] tcp://127.0.0.1:44327 READY
2021-05-14T11:06:05.4668533Z 2021/05/14 11:04:05 [ERROR] [fs.WatchFile:/var/lib/networkservicemesh/config/excluded_prefixes.yaml] can not create directory: mkdir /var/lib/networkservicemesh: permission denied
2021-05-14T11:06:05.4669482Z 2021/05/14 11:04:05 [DEBUG] tcp://127.0.0.1:36087 READY
2021-05-14T11:06:05.4669960Z 2021/05/14 11:04:05 [DEBUG] tcp://127.0.0.1:44327 READY
2021-05-14T11:06:05.4670428Z 2021/05/14 11:04:05 [DEBUG] tcp://127.0.0.1:43643 READY
2021-05-14T11:06:05.4670925Z 2021/05/14 11:04:05 [INFO]  Stop serve: tcp://127.0.0.1:43643
2021-05-14T11:06:05.4671807Z 2021/05/14 11:04:05 [DEBUG] START RESTORE: nsmgr-beb8804b-e1d0-4986-8262-d311316ee98b
2021-05-14T11:06:05.4672847Z 2021/05/14 11:04:05 [DEBUG] END RESTORE: nsmgr-beb8804b-e1d0-4986-8262-d311316ee98b
2021-05-14T11:06:05.4673558Z 2021/05/14 11:04:05 [DEBUG] tcp://127.0.0.1:43643 CONNECTING
2021-05-14T11:06:05.4674096Z 2021/05/14 11:04:05 [DEBUG] tcp://127.0.0.1:43643 TRANSIENT_FAILURE
2021-05-14T11:06:05.4675205Z 2021/05/14 11:04:05 [INFO]  Started listening endpoint 73734e74-6408-4ff5-a6a0-0ecab6eed056-final-endpoint-2 on tcp://127.0.0.1:40767
2021-05-14T11:06:05.4676081Z 2021/05/14 11:04:15 [INFO]  Stop serve: tcp://127.0.0.1:40767
2021-05-14T11:06:05.4676585Z 2021/05/14 11:04:15 [INFO]  Stop serve: tcp://127.0.0.1:36087
2021-05-14T11:06:05.4677098Z 2021/05/14 11:04:15 [INFO]  Stop serve: tcp://127.0.0.1:44327
2021-05-14T11:06:05.4677602Z 2021/05/14 11:04:15 [INFO]  Stop serve: tcp://127.0.0.1:37911
2021-05-14T11:06:05.4678105Z 2021/05/14 11:04:15 [INFO]  Stop serve: tcp://127.0.0.1:45251
2021-05-14T11:06:05.4678602Z 2021/05/14 11:04:15 [INFO]  Stop serve: tcp://127.0.0.1:36973
2021-05-14T11:06:05.4679565Z 2021/05/14 11:04:15 [DEBUG] START RESTORE: forwarder-722ddefe-9a47-4333-ad16-668898ca774b
2021-05-14T11:06:05.4680779Z 2021/05/14 11:04:15 [DEBUG] END RESTORE: forwarder-722ddefe-9a47-4333-ad16-668898ca774b
2021-05-14T11:06:05.4681917Z 2021/05/14 11:04:15 [DEBUG] START RESTORE: nsmgr-beb8804b-e1d0-4986-8262-d311316ee98b
2021-05-14T11:06:05.4682966Z 2021/05/14 11:04:15 [DEBUG] END RESTORE: nsmgr-beb8804b-e1d0-4986-8262-d311316ee98b
2021-05-14T11:06:05.4684056Z 2021/05/14 11:04:15 [DEBUG] START RESTORE: nsmgr-263cb81b-eebf-4eef-b3bf-87c8f6d111ee
2021-05-14T11:06:05.4685298Z 2021/05/14 11:04:15 [DEBUG] END RESTORE: nsmgr-263cb81b-eebf-4eef-b3bf-87c8f6d111ee
2021-05-14T11:06:05.4686454Z 2021/05/14 11:04:15 [DEBUG] START RESTORE: nsmgr-263cb81b-eebf-4eef-b3bf-87c8f6d111ee
2021-05-14T11:06:05.4687604Z 2021/05/14 11:04:15 [DEBUG] END RESTORE: nsmgr-263cb81b-eebf-4eef-b3bf-87c8f6d111ee
2021-05-14T11:06:05.4688827Z 2021/05/14 11:04:15 [DEBUG] START RESTORE: forwarder-bebfdb01-0f58-4810-8497-eda567dbcbfc
2021-05-14T11:06:05.4690188Z 2021/05/14 11:04:15 [DEBUG] END RESTORE: forwarder-bebfdb01-0f58-4810-8497-eda567dbcbfc
2021-05-14T11:06:05.4691198Z 2021/05/14 11:04:15 [DEBUG] --- END: TestNSMGR_HealEndpoint/Remote_New
2021-05-14T11:06:05.4691899Z --- FAIL: TestNSMGR_HealEndpoint (16.61s)
2021-05-14T11:06:05.4692610Z     --- FAIL: TestNSMGR_HealEndpoint/Remote_New (10.11s)
2021-05-14T11:06:05.4700668Z         heal_test.go:142: 
2021-05-14T11:06:05.4701149Z             	Error Trace:	heal_test.go:142
2021-05-14T11:06:05.4701603Z             	            				heal_test.go:77
2021-05-14T11:06:05.4702088Z             	Error:      	Condition never satisfied
2021-05-14T11:06:05.4702669Z             	Test:       	TestNSMGR_HealEndpoint/Remote_New
```

**Comment by denis-tingaikin** [comment]
@Bolodya1997 Could you please attach logs?

**Comment by Bolodya1997** [comment]
> @Bolodya1997 Could you please attach logs?

Yep, updated the description.

**Comment by denis-tingaikin** [comment]
> If monitor gRPC stream fails to start, gRPC connection should be closed and all Requests should return an error.

Why is this expected?

**Comment by denis-tingaikin** [comment]
Should we consider monitoring healing?

**Comment by Bolodya1997** [comment]
> Why is this expected?
> Should we consider monitoring healing?

It is actually the way how we are healing monitor now - closing gRPC connection and starting a new one.

**Comment by edwarnicke** [comment]
@Bolodya1997 Do we know why we are seeing cases where:
> 1. Sometimes we have gRPC stream closed, but gRPC connection still working and passing Requests.
> 2. Sometimes gRPC stream fails to start, but Requests return success.

That seems unusual, I'd love to understand the *why* there.

**Comment by Bolodya1997** [comment]
Being trying to reproduce such gRPC behavior locally on simple test, but failed.
Probably I was wrong and initial issue wasn't caused by gRPC, but only by race bugs, or probably there is something else on CI that should be additionally tested and doesn't reveal on local tests.
So I guess this should be additionally retested on CI in some close future.

**Comment by Bolodya1997** [comment]
> Probably I was wrong and initial issue wasn't caused by gRPC, but only by race bugs, or probably there is something else on CI that should be additionally tested and doesn't reveal on local tests.

Merging just #926 doesn't fix the issue. So we still need #927 and it doesn't look like this issue caused only by race bugs.

**Comment by d-uzlov** [comment]
While making a test for client healing for https://github.com/networkservicemesh/sdk/pull/945, I accidentally stumbled upon this bug.

In [this test](https://github.com/d-uzlov/sdk/blob/7df47a28bab764e98105a691252843779a604a17/pkg/networkservice/chains/client/client_heal_test.go#L37) I cancel the context, and endpoint server and the connection to it die. However, connections monitor dies before we detect that connection dies, so when heal happens, we don't delete the old connect client.
Then the endpoint server is restarted on the same URL, and when we do further requests, they successfully reach the endpoint, but the connections monitor is dead.

[Related PR](https://github.com/networkservicemesh/sdk/pull/927) seem to fix the issue in this test.

**Comment by Bolodya1997** [comment]
Finally closed with #927 merged.
